### PR TITLE
[AC-1608] Send offboarding survey response to Stripe on subscription cancellation

### PR DESCRIPTION
**Pull request by amorask-bitwarden**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

The purpose of this PR is to update the subscription cancellation endpoints for both Organizations and Users to take in the results of an offboarding survey presented to users on the client-side and then pass that information to Stripe when cancelling the subscription and a feature flag is on.

## Important Updates

* **SubscriptionCancellationRequestModel.cs:** A request model representing the results of the offboarding survey.
* **OrganizationsController.cs:** Uses a feature flag to determine whether to call the new `CancelSubscriptionCommand` or the existing `OrganizationService.CancelSubscriptionAsync` method.
* **AccountsController.cs:** Uses a feature flag to determine whether to call the new `CancelSubscriptionCommand` or the existing `UserService.CancelPremiumAsync` method.
* **Organization.cs:** Added some `ISubscriber` helper methods to assist in processing.
* **User.cs:** Added some `ISubscriber` helper methods to assist in processing.
* **ISubscriber.cs:** Added some helper methods to assist in processing.
* **CancelSubscriptionCommand.cs:** New command to continue the process of removing methods from the bloated `StripePaymentService`. Handles the new logic behind the FF.
* **Constants.cs:** Added the "AC-1607_present-user-offboarding-survey" feature flag.
* **Startup.cs:** Had to register the billing commands in Identity because one is injected into the `UserService`.

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

**Review by cyprain-okeke — COMMENTED** [review]
Looks Good. Thanks for adding a lot unit test

File: src/Api/AdminConsole/Controllers/OrganizationsController.cs [hunk 1]
```diff
@@ -444,15 +455,44 @@ public async Task PostVerifyBank(string id, [FromBody] OrganizationVerifyBankReq
 
     [HttpPost("{id}/cancel")]
     [SelfHosted(NotSelfHostedOnly = true)]
-    public async Task PostCancel(string id)
+    public async Task PostCancel(string id, [FromBody] SubscriptionCancellationRequestModel request)
```
Comment (r-tome) [inline src/Api/AdminConsole/Controllers/OrganizationsController.cs:458]:
```suggestion
    public async Task PostCancel(Guid id, [FromBody] SubscriptionCancellationRequestModel request)
```

File: src/Api/AdminConsole/Controllers/OrganizationsController.cs [hunk 2]
```diff
@@ -444,15 +455,44 @@ public async Task PostVerifyBank(string id, [FromBody] OrganizationVerifyBankReq
...
     {
         var orgIdGuid = new Guid(id);
+
         if (!await _currentContext.EditSubscription(orgIdGuid))
         {
             throw new NotFoundException();
         }
 
-        await _organizationService.CancelSubscriptionAsync(orgIdGuid);
+        if (_featureService.IsEnabled(FeatureFlagKeys.AC1607_PresentUsersWithOffboardingSurvey))
```
Comment (r-tome) [inline src/Api/AdminConsole/Controllers/OrganizationsController.cs:467]: ⛏️ I'd prefer if the feature flag was set to a var, it'd be easier to read.

File: src/Api/Auth/Controllers/AccountsController.cs [hunk 3]
```diff
@@ -805,15 +819,36 @@ public async Task PostLicense(LicenseRequestModel model)
...
     {
         var user = await _userService.GetUserByPrincipalAsync(User);
+
         if (user == null)
         {
             throw new UnauthorizedAccessException();
         }
 
-        await _userService.CancelPremiumAsync(user);
+        if (_featureService.IsEnabled(FeatureFlagKeys.AC1607_PresentUsersWithOffboardingSurvey))
```
Comment (r-tome) [inline src/Api/Auth/Controllers/AccountsController.cs:831]: ⛏️ Same thing here, I'd put this value on a var

File: src/Core/Billing/Commands/Implementations/CancelSubscriptionCommand.cs [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    public async Task CancelSubscription(
+        ISubscriber subscriber,
+        OffboardingSurveyResponse offboardingSurveyResponse)
+    {
+        var subscription = await GetSubscriptionAsync(subscriber);
+
+        if (IsInactive(subscription))
+        {
+            _logger.LogWarning("Cannot cancel subscription ({ID}) that's already inactive.", subscription.Id);
+            return;
```
Comment (r-tome) [inline src/Core/Billing/Commands/Implementations/CancelSubscriptionCommand.cs:44]: If we don't throw an `Exception` here then on the `Controller` its considered as a success and then it sends the event to `ReferenceEventService`, correct?

File: src/Core/Billing/Commands/Implementations/CancelSubscriptionCommand.cs [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    {
+        _logger = logger;
+        _stripeAdapter = stripeAdapter;
+    }
+
+    public async Task CancelSubscription(
+        ISubscriber subscriber,
+        OffboardingSurveyResponse offboardingSurveyResponse)
+    {
+        var subscription = await GetSubscriptionAsync(subscriber);
```
Comment (r-tome) [inline src/Core/Billing/Commands/Implementations/CancelSubscriptionCommand.cs:39]: To fully implement the CQRS pattern this should be an input parameter

File: src/Core/Billing/Commands/Implementations/CancelSubscriptionCommand.cs [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        {
+            options.CancellationDetails.Feedback = offboardingSurveyResponse.Reason;
+        }
+
+        await _stripeAdapter.SubscriptionCancelAsync(subscriptionId, options);
+    }
+
+    private static GatewayException ContactSupport() => new("Something went wrong when trying to cancel your subscription. Please contact support.");
+
+    private async Task<Subscription> GetSubscriptionAsync(ISubscriber subscriber)
```
Comment (r-tome) [inline src/Core/Billing/Commands/Implementations/CancelSubscriptionCommand.cs:92]: This method could be a query command `IGetSubscriptionCommand`

**Review by r-tome — CHANGES_REQUESTED** [review]

File: src/Core/Billing/Commands/Implementations/CancelSubscriptionCommand.cs (hunk 4 above)
Comment (amorask-bitwarden) [reply src/Core/Billing/Commands/Implementations/CancelSubscriptionCommand.cs:44]: @r-tome Good catch; I think this should throw an exception.

File: src/Core/Billing/Queries/IGetSubscriptionQuery.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
+﻿using Bit.Core.Entities;
+using Stripe;
+
+namespace Bit.Core.Billing.Queries;
+
+public interface IGetSubscriptionQuery
+{
+    /// <summary>
+    ///
```
Comment (r-tome) [inline src/Core/Billing/Queries/IGetSubscriptionQuery.cs:9]: Summary missing

**Review by r-tome — APPROVED** [review]
Great work! Thanks for making the changes I suggested

File: src/Core/Billing/Queries/IGetSubscriptionQuery.cs (hunk 7 above)
Comment (amorask-bitwarden) [reply src/Core/Billing/Queries/IGetSubscriptionQuery.cs]: @r-tome Whoops! Nice catch. I just added this - thanks.

**Review by r-tome — DISMISSED** [review]

**Review by cyprain-okeke — APPROVED** [review]

**Review by r-tome — DISMISSED** [review]

**Review by ike-kottlowski — DISMISSED** [review]
Looks straight forward. Good work.

**Comment by amorask-bitwarden** [comment]
@cyprain-okeke @r-tome @ike-kottlowski Apologies, can I ask you three to re-stamp when you have a second? I had to resolve some small merge conflicts.

**Review by r-tome — APPROVED** [review]

**Review by cyprain-okeke — APPROVED** [review]

**Review by ike-kottlowski — APPROVED** [review]

**Pull request merged**